couchbase/src/main/scala/akka/stream/alpakka/couchbase/impl/CouchbaseSessionImpl.scala

```diff
@@ -4,179 +4,79 @@
-      .map(_.content(): Long)(ExecutionContext.parasitic)
-
-  def counter(id: String, delta: Long, initial: Long, writeSettings: CouchbaseWriteSettings): Future[Long] =
-    singleObservableToFuture(asyncBucket.counter(id,
-                                                 delta,
-                                                 initial,
-                                                 writeSettings.persistTo,
-                                                 writeSettings.replicateTo,
-                                                 writeSettings.timeout.toMillis,
-                                                 TimeUnit.MILLISECONDS),
-                             id)
-      .map(_.content(): Long)(ExecutionContext.parasitic)
+  override def underlying: AsyncBucket = cluster.bucket(bucketName)
+
+  override def streamedQuery(query: String): Source[JsonObject, NotUsed] = {
+    Source.fromIterator(
+      () =>
+        cluster
+          .query(query)
+          .get()
+          .rowsAsObject()
+          .iterator()
+          .asScala
+    )
+  }
+
+  override def streamedQuery(query: String, queryOptions: QueryOptions): Source[JsonObject, NotUsed] =
+    Source.fromIterator(
+      () =>
+        cluster
+          .query(query, queryOptions)
+          .get()
+          .rowsAsObject()
+          .iterator()
+          .asScala
+    )
 
   def close(): Future[Done] =
-    if (!asyncBucket.isClosed) {
-      singleObservableToFuture(asyncBucket.close(), "close")
-        .flatMap { _ =>
-          cluster match {
-            case Some(cluster) =>
-              singleObservableToFuture(cluster.disconnect(), "close").map(_ => Done)(ExecutionContext.global)
-            case None => Future.successful(Done)
-          }
-        }(ExecutionContext.global)
-    } else {
-      Future.successful(Done)
-    }
-
-  override def toString: String = s"CouchbaseSession(${asyncBucket.name()})"
+    Future.successful(Done)
 
-  override def createIndex(indexName: String, ignoreIfExist: Boolean, fields: AnyRef*): Future[Boolean] =
-    singleObservableToFuture(
-      asyncBucket
-        .bucketManager()
-        .flatMap(
-          func1Observable[AsyncBucketManager, Boolean](
-            _.createN1qlIndex(indexName, ignoreIfExist, false, fields: _*)
-              .map(func1(Boolean.unbox))
-          )
-        ),
-      s"Create index: $indexName"
-    )
-
-  override def listIndexes(): Source[IndexInfo, NotUsed] =
-    Source.fromPublisher(
-      RxReactiveStreams.toPublisher(
-        asyncBucket
-          .bucketManager()
-          .flatMap(
-            func1Observable((abm: AsyncBucketManager) => abm.listN1qlIndexes())
-          )
-      )
-    )
+  override def toString: String = s"CouchbaseSession(${underlying.name()})"
 
+  private def getSingleResult(gr: QueryResult): Option[JsonObject] = {
+    val rows = gr.rowsAsObject()
+    if (rows.isEmpty) {
+      return Option.empty
+    }
+    Option.apply(rows.iterator.next)
```

Comment:
Should it not also check that there is only one object and fail if there are more? Or is it perhaps fine that `singleResponseQuery` also can be used as a `firstResponseQuery`?

Reply:
I don't think it should check it here as that would happen only if the original query returned multiple results and developer explicitly showed the desire to receive only the first result (by choosing methods that use this function). I've documented that behavior. The performance in cases when users don't use `LIMIT 1` and query returns more than one row can be improved by wrapping the passed query as a sub-query inside a query with `LIMIT 1`, but I think that optimization is such a fringe case that it can be implemented separately.